### PR TITLE
UGENE-7720: Fix scrolling for short sequences

### PR DESCRIPTION
**Pull request by ichebyki**
[Fix scrolling for short sequences](https://github.com/ugeneunipro/ugene/commit/6b420912c3d0e19dcd5ad35bb33a6c78e992088b)

### CODE REVIEW

**Review by yalgaer — CHANGES_REQUESTED** [review]
Looks like this PR includes multiple unrelated changes. Please update.

**Comment by ichebyki** [comment]
Sure

пт, 18 нояб. 2022 г., 23:42 Yuliya Algaer ***@***.***>:

**Comment by ichebyki** [comment]
> Looks like this PR includes multiple unrelated changes. Please update.

Done

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp [hunk 1]
```diff
@@ -177,11 +177,58 @@ int MultilineScrollController::getViewHeight() {
...
+            break;
+        }
+        if ((firstBase + 1) >= alignmentLen) {
+            needUpdate = true;
+            break;
+        }
+        prevFistBase = firstBase;
+    }
+    if (needUpdate) {
+        const int length = ui->getSequenceAreaBaseLen(0);
```
Comment (yalgaer) [inline src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: nit: do not use const for locals (you use it inconsistently across the patch and UGENE code (99%) does not use it at all

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp [hunk 2]
```diff
@@ -177,11 +177,58 @@ int MultilineScrollController::getViewHeight() {
...
+    bool needUpdate = false;
+    int childrenCount = ui->getChildrenCount();
+    int alignmentLen = maEditor->getAlignmentLen();
+    for (int i = 0; i < childrenCount; i++) {
+        firstBase = ui->getUI(i)->getScrollController()->getFirstVisibleBase(false);
+        if (firstBase == prevFistBase) {
+            needUpdate = true;
+            break;
+        }
+        if ((firstBase + 1) >= alignmentLen) {
```
Comment (yalgaer) [inline src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: All these braces with +1 are not needed in the patch

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp [hunk 3]
```diff
@@ -177,11 +177,58 @@ int MultilineScrollController::getViewHeight() {
...
 }
 
+bool MultilineScrollController::checkBoundary() {
+    int firstBase = -1;
+    int prevFistBase = -1;
+    bool needUpdate = false;
+    int childrenCount = ui->getChildrenCount();
+    int alignmentLen = maEditor->getAlignmentLen();
+    for (int i = 0; i < childrenCount; i++) {
+        firstBase = ui->getUI(i)->getScrollController()->getFirstVisibleBase(false);
```
Comment (yalgaer) [inline src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: getUI(i) can return nullptr right?

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp [hunk 4]
```diff
@@ -177,11 +177,58 @@ int MultilineScrollController::getViewHeight() {
...
+        prevFistBase = firstBase;
+    }
+    if (needUpdate) {
+        const int length = ui->getSequenceAreaBaseLen(0);
+        if ((firstBase + 1) >= alignmentLen || firstBase == prevFistBase) {
+            firstBase = firstBase == 0
+                            ? 0
+                            : alignmentLen / length * length -
+                                  length * (childrenCount - (alignmentLen % length > 0 ? 1 : 0));
+            if (firstBase < 0) {
```
Comment (yalgaer) [inline src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: nit: can be 1-liner: qMax

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp [hunk 5]
```diff
@@ -177,11 +177,58 @@ int MultilineScrollController::getViewHeight() {
...
+            prevFistBase = -1;
+            for (int i = 0; i < childrenCount; i++) {
+                if (firstBase == prevFistBase ||
+                    ((firstBase + 1) >= alignmentLen && alignmentLen > 1) ||
+                    ((firstBase + 1) >= alignmentLen && alignmentLen == 1 && prevFistBase != -1)) {
+                    ui->getUI(i)->setHidden(true);
+                } else {
+                    ui->getUI(i)->setHidden(false);
+                    QSignalBlocker signalBlocker(ui->getUI(i)->getScrollController());
+                    Q_UNUSED(signalBlocker);
```
Comment (yalgaer) [inline src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: Q_UNUSED is not needed here

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp [hunk 6]
```diff
@@ -308,9 +355,9 @@ void MultilineScrollController::centerViewRow(QPoint maPoint) {
     vScrollBar->setValue(newScreenYOffset);
 }
 
-void MultilineScrollController::centerPoint(const QPoint& maPoint) {
-    centerBase(maPoint.x());
-    centerViewRow(maPoint);
+void MultilineScrollController::_centerPoint(const QPoint& maPoint) {
+    _centerBase(maPoint.x());
```
Comment (yalgaer) [inline src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: do not use underscore for method names (this style is not used by UGENE).

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.h [hunk 7]
```diff
@@ -138,11 +136,14 @@ public slots:
...
+
     // TODO:ichebyki
     // Deprecated, will be removed
-    void centerBase(int baseNumber);
-    void centerPoint(const QPoint& maPoint);
-    void setCenterVisibleBase(int firstVisibleBase);
+    void _centerBase(int baseNumber);
+    void _centerPoint(const QPoint& maPoint);
+    void _setCenterVisibleBase(int firstVisibleBase);
+    void _centerViewRow(QPoint maPoint);
```
Comment (yalgaer) [inline src/corelibs/U2View/src/ov_msa/MultilineScrollController.h]: Add comments for all new methods

Comment (yalgaer) [inline src/corelibs/U2View/src/ov_msa/MultilineScrollController.h]: Use const T& for all non-primitive types

**Review by yalgaer — CHANGES_REQUESTED** [review]

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp (hunk 1 above)
Comment (ichebyki) [reply src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: fixed

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp (hunk 3 above)
Comment (ichebyki) [reply src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: Fixed

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp (hunk 4 above)
Comment (ichebyki) [reply src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: Fixed

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp (hunk 5 above)
Comment (ichebyki) [reply src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: Fixed

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp (hunk 6 above)
Comment (ichebyki) [reply src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: The methods are deprecated and will be removed soon
Underscore is used as additional reminder

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.h (hunk 7 above)
Comment (ichebyki) [reply src/corelibs/U2View/src/ov_msa/MultilineScrollController.h]: Fixed

Comment (ichebyki) [reply src/corelibs/U2View/src/ov_msa/MultilineScrollController.h]: Fixed

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp [hunk 8]
```diff
@@ -177,11 +177,59 @@ int MultilineScrollController::getViewHeight() {
...
 
+bool MultilineScrollController::checkBoundary() {
+    int firstBase = -1;
+    int prevFistBase = -1;
+    bool needUpdate = false;
+    int childrenCount = ui->getChildrenCount();
+    int alignmentLen = maEditor->getAlignmentLen();
+    for (int i = 0; i < childrenCount; i++) {
+        MaEditorWgt* wgt = ui->getUI(i);
+        if (wgt == nullptr) {
```
Comment (yalgaer) [inline src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: nit: use `CHECK(wgt != nullptr)` of you expect nullptr or `SAFE_POINT` if you do not.
In this case it looks like it must be SAFEPOINT

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp [hunk 9]
```diff
@@ -177,11 +177,59 @@ int MultilineScrollController::getViewHeight() {
...
+        prevFistBase = firstBase;
+    }
+    if (needUpdate) {
+        int length = ui->getSequenceAreaBaseLen(0);
+        if ((firstBase + 1) >= alignmentLen || firstBase == prevFistBase) {
+            firstBase = qMax(0, firstBase == 0 ? 0 : alignmentLen / length * length - length * (childrenCount - (alignmentLen % length > 0 ? 1 : 0)));
+            prevFistBase = -1;
+            for (int i = 0; i < childrenCount; i++) {
+                MaEditorWgt* wgt = ui->getUI(i);
+                if (wgt == nullptr) {
```
Comment (yalgaer) [inline src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: nit: same here. Use SAFE_POINT?

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp [hunk 10]
```diff
@@ -534,6 +583,9 @@ void MultilineScrollController::updateChildrenScrollBarsPeivate() {
         if (i == 0) {
             val = hbar->value();
         }
+
+        QSignalBlocker signalBlocker(ui->getUI(i)->getScrollController());
+        Q_UNUSED(signalBlocker);
```
Comment (yalgaer) [inline src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: 1. Q_UNUSED is not needed here
2. getUi(i) can return nullptr.

**Review by yalgaer — CHANGES_REQUESTED** [review]
See unresolved comments above

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp (hunk 2 above)
Comment (ichebyki) [reply src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: fixed

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp (hunk 8 above)
Comment (ichebyki) [reply src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: fixed

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp (hunk 9 above)
Comment (ichebyki) [reply src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: fixed

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp (hunk 10 above)
Comment (ichebyki) [reply src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: fixed

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp (hunk 6 above)
Comment (ichebyki) [reply src/corelibs/U2View/src/ov_msa/MultilineScrollController.cpp]: all these methods was removed

File: src/corelibs/U2View/src/ov_msa/MultilineScrollController.h (hunk 7 above)
Comment (ichebyki) [reply src/corelibs/U2View/src/ov_msa/MultilineScrollController.h]: all these methods was removed

Comment (ichebyki) [reply src/corelibs/U2View/src/ov_msa/MultilineScrollController.h]: all these methods was removed

**Comment by yalgaer** [comment]
I tried to open amino_ext.aln (from tests) in a small window and got this screenshot:

image


To reproduce:
1) Reduce UGENE window size to match the screenshot
2) Open amino_ext.aln: the view is rendered correctly: 2 lines
3) Reduce the view size with a slider: only 2 lines are rendered and the third line is not visible (see the screenshot)

**Comment by ichebyki** [comment]
> 2\. amino_ext.aln

fixed

**Review by yalgaer — COMMENTED** [review]
The code looks good.

Is it possible to write a GUI test to avoid such issues in the future?

**Comment by ichebyki** [comment]
> The code looks good.
> 
> Is it possible to write a GUI test to avoid such issues in the future?

sure

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp [hunk 11]
```diff
@@ -3489,5 +3488,89 @@ GUI_TEST_CLASS_DEFINITION(test_7715) {
...
+    GTMouseDriver::press();
+    GTMouseDriver::moveTo(p);
+    GTMouseDriver::release();
+
+    GTThread::waitForMainThread();
+    GTUtilsMsaEditor::setMultilineMode(os, true);
+
+    auto count = GTUtilsMsaEditor::getEditor(os)->getUI()->getChildrenCount();
+    bool allAreVisible = true;
+    for (uint i = 0; i < count; i++) {
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: Do not use unsigned types unless it is required by some external library.
Unsigned types introduce warnings and unexpected behaviors when used in the same expressions with signed types.

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp [hunk 12]
```diff
@@ -3489,5 +3488,89 @@ GUI_TEST_CLASS_DEFINITION(test_7715) {
...
+    CHECK_SET_ERR(activeWindow != nullptr, "Sequence widget exists");
+
+    wgt = GTUtilsMSAEditorSequenceArea::getSequenceArea(os, 0);
+    CHECK_SET_ERR(wgt != nullptr, "Sequence area exists");
+
+    splitter = GTWidget::findSplitter(os, "name_and_sequence_areas_splitter", activeWindow);
+    CHECK_SET_ERR(splitter != nullptr, "Splitter exists");
+
+    handle = splitter->handle(1);
+    CHECK_SET_ERR(handle != nullptr, "MSA Splitter handle is not NULL");
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: Low level/non-documented code that should be extracted into utility method and reused by other tests.

If you do not want to create an utility method that moves MSA splitter use other options , for example move project view splitter (search for `GTMouseDriver::dragAndDrop(splitterCenter, splitterCenter + delta);`

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp [hunk 13]
```diff
@@ -3489,5 +3488,89 @@ GUI_TEST_CLASS_DEFINITION(test_7715) {
...
+
+    allAreVisible = true;
+    for (uint i = 0; i < count; i++) {
+        auto w = GTUtilsMsaEditor::getEditor(os)->getUI()->getUI(i);
+        CHECK_SET_ERR(w != nullptr, "Sequence widget exists");
+        allAreVisible = allAreVisible && w->isVisible();
+    }
+    CHECK_SET_ERR(!allAreVisible, "Some line of the multiline view must be hidden");
+
+    GTUtilsMsaEditor::setMultilineMode(os, false);
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: This op is not needed

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp [hunk 14]
```diff
@@ -3489,5 +3488,89 @@ GUI_TEST_CLASS_DEFINITION(test_7715) {
...
+    GTFileDialog::openFile(os, testDir + "_common_data/realign_sequences_in_alignment/", "amino_ext.aln");
+    GTUtilsTaskTreeView::waitTaskFinished(os);
+
+    GTUtilsMsaEditor::setMultilineMode(os, false);
+
+    auto activeWindow = GTUtilsMsaEditor::getEditor(os)->getUI()->getUI(0);
+    CHECK_SET_ERR(activeWindow != nullptr, "Sequence widget is visible");
+
+    auto wgt = GTUtilsMSAEditorSequenceArea::getSequenceArea(os, 0);
+    CHECK_SET_ERR(wgt != nullptr, "Sequence area is visible");
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: nit: The method `GTUtilsMSAEditorSequenceArea::getSequenceArea` is designed incorrectly - it should never return nulls by default. So the called won't be needed to check and all usages in the old (non-multiline mode) will stay safe (old usages do not check).

**Review by yalgaer — CHANGES_REQUESTED** [review]

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp (hunk 11 above)
Comment (ichebyki) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: <MaEditorMultilineWgt.h>:

    // Return lines count in multiline widget
    uint getChildrenCount() const {
        return uiChildCount;
    }

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp (hunk 12 above)
Comment (ichebyki) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: fixed

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp (hunk 11 above)
Comment (ichebyki) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: already fixed

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp (hunk 13 above)
Comment (ichebyki) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: fixed

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp (hunk 14 above)
Comment (ichebyki) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: fixed

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp (hunk 11 above)
Comment (yalgaer) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: I had the same comment to the original big patch: all unsigned types you introduced should go away. At the time I'm writing this comment the issue is not fixed

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp [hunk 15]
```diff
@@ -3489,5 +3488,72 @@ GUI_TEST_CLASS_DEFINITION(test_7715) {
...
+    auto activeWindow = GTUtilsMsaEditor::getEditor(os)->getUI()->getUI(0);
+    CHECK_SET_ERR(activeWindow != nullptr, "Sequence widget is not visible");
+
+    auto splitter = GTWidget::findSplitter(os, "name_and_sequence_areas_splitter", activeWindow);
+    CHECK_SET_ERR(splitter != nullptr, "Splitter is not visible");
+
+    QSplitterHandle* handle = splitter->handle(1);
+    CHECK_SET_ERR(handle != nullptr, "MSA Splitter handle is NULL");
+
+    GTWidget::click(os, handle);
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: Do you really need this click here? May be `move` or just getting a center coordinate of the widget for the later dragAndDrop should be enough?

Comment (ichebyki) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: removed

**Review by yalgaer — COMMENTED** [review]
Just an idea: If you have problems with the test & slider you can try another approach. Try opening/closing Project view & Options panel in order to reduce or enlarge MSA sequence view

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp (hunk 11 above)
Comment (ichebyki) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: Let's change uint -> int in separate issue/PR ?
Too many places to change - the this PR will be messed

Comment (yalgaer) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: Please fix only new uints introduced in this PR. AFAIU all other uints in UGENE were already cleaned up.

**Comment by yalgaer** [comment]
I merged with master/fixed uints to ints and re-run GUI tests.
Will approve once they finish with no issues.

**Pull request merged**